gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/ExpressionContext.java

```diff
@@ -454,17 +338,256 @@ private void getSourceToFilterExpressionMap(
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    Map.Entry::getValue,
+                    (v1, v2) -> {
+                      Set<String> mergedSet = new HashSet<>(v1);
+                      mergedSet.addAll(v2);
+                      return mergedSet;
+                    },
+                    HashMap::new));
+
+    // A weird case, if there are no filters and order bys
+    if (sourceToFilterAttributeMap.isEmpty() && sourceToOrderByAttributeMap.isEmpty()) {
+      return Collections.emptySet();
+    }
+
+    Map<String, Set<String>> filterAttributeToSourcesMap =
+        expressionContext.getFilterAttributeToSourceMap();
+    Map<String, Set<String>> orderByAttributeToSourceMap =
+        buildAttributeToSourcesMap(sourceToOrderByAttributeMap);
+
+    return getIntersectingSourceSets(filterAttributeToSourcesMap, orderByAttributeToSourceMap);
+  }
+
+  /**
+   * Given a source to attributes, builds an attribute to sources map. Basically, a reverse map of
+   * the map provided as input
+   *
+   * <p>Example:
+   *
+   * <p>("QS" -> API.id, "QS" -> API.name, "EDS" -> API.id) =>
+   *
+   * <p>("API.id" -> ["QS", "EDS"], "API.name" -> "QS")
+   */
+  private static Map<String, Set<String>> buildAttributeToSourcesMap(
+      Map<String, Set<String>> sourcesToAttributeMap) {
+    Map<String, Set<String>> attributeToSourcesMap = new HashMap<>();
+    for (Map.Entry<String, Set<String>> entry : sourcesToAttributeMap.entrySet()) {
+      String source = entry.getKey();
+      for (String attribute : entry.getValue()) {
+        attributeToSourcesMap.computeIfAbsent(attribute, k -> new HashSet<>()).add(source);
+      }
+    }
+    return Collections.unmodifiableMap(attributeToSourcesMap);
+  }
+
+  /**
+   * Computes intersecting source sets from 2 attribute to sources map i.e. computes intersection
+   * source sets across all attributes from the map
+   *
+   * <pre>
+   * Examples:
+   *
+   * 1.
+   * ("API.id" -> ["EDS", "QS"], "API.name" -> ["QS", "EDS"])
+   * ("API.id" -> ["EDS", "QS"], "API.discoveryState" -> ["EDS"])
+   *
+   * The intersecting source set across all the attributes would be ["EDS"]
+   *
+   * 2.
+   * ("API.id" -> ["EDS", "QS"], "API.name" -> ["QS", "EDS"])
+   * ("API.id" -> ["EDS", "QS"], "API.discoveryState" -> ["EDS", "QS"])
+   *
+   * The intersecting source set across all the attributes would be ["EDS", "QS"]
+   *
+   * 3.
+   * ("API.id" -> ["EDS"], "API.name" -> ["EDS"])
+   * ("API.id" -> ["EDS"], "API.discoveryState" -> ["QS"])
+   *
+   * The intersecting source set across all the attributes would be []
+   *
+   * </pre>
+   */
+  private static Set<String> getIntersectingSourceSets(
+      Map<String, Set<String>> attributeToSourcesMapFirst,
+      Map<String, Set<String>> attributeToSourcesMapSecond) {
+    if (attributeToSourcesMapFirst.isEmpty() && attributeToSourcesMapSecond.isEmpty()) {
```

Comment:
we want the intersection of values, can we do
```
Stream.concat(
    attributeToSourcesMapFirst.values().stream(), 
    attributeToSourcesMapSecond.values().stream())
.reduce(Sets::intersection)
.orElse(Collections.emptySet());
```

Reply:
An existing code, being refactored to a different class